source/slang/slang-ast-modifier.h

```diff
@@ -1038,9 +1043,12 @@ class NumThreadsAttribute : public Attribute
     //
     // TODO: These should be accessors that use the
     // ordinary `args` list, rather than side data.
-    IntVal* x;
-    IntVal* y;
-    IntVal* z;
+    IntVal* extents[3];
+
+    // References to specialization constants, for defining the number of
+    // threads with them. If set, the corresponding axis is set to nullptr
+    // above.
+    DeclRef<VarDeclBase> specConstExtents[3];
```

Comment:
It is unfortunate that we need to use this side-band mechanism to handle the specialization constant case, but I also cannot think of any simpler way without complicating the `IntVal` hierarchy and the constant folding logic.

Reply:
Yeah, I don't like it either. Especially the `axisIsSpecConstId` is nasty in the GLSL version, but since GLSL can legally reference undeclared specialization constant IDs, there's probably not much else that can be done.